tests/test_client_functional.py

```diff
@@ -2927,6 +2927,73 @@ async def test_creds_in_auth_and_url() -> None:
         await session.close()
 
 
+@pytest.mark.parametrize(
+    ["check_redirect_auth_precedence"],
+    ([False], [True]),
+    ids=("redirect auth clash allowed", "redirect auth precedence"),
+)
+async def test_creds_in_auth_and_redirect_url(
+    create_server_for_url_and_handler: Callable[[URL, Handler], Awaitable[TestServer]],
+    check_redirect_auth_precedence: bool,
```

Comment:
I see that this argument is only used for optional assertion, so I assume what the behavior of both test variants is exactly the same. I'd expect some configuration difference between them..

Reply:
The (maybe wrong) rationale behind it is that you can break this test in 2 ways: you can either break the part that allows the clash on redirect, which will break both variants, or break the part that overrides the auth, which will only break the second variant. I figured this would make it easier to pinpoint the issue if someone breaks this somewhere down the line. Is there a better way to achieve that?

Reply:
The tests should usually orchestrate for things they rely on to happen so that they check them. Are you saying it's somehow random?

Reply:
Not at all. I'm saying that if someone breaks this behaviour

https://github.com/aio-libs/aiohttp/blob/549279a36a815468a828a98c3375c1f7b2949285/aiohttp/client.py#L560-L563

the clash-allowed and precedence-enforced test variants will both fail. And if someone breaks this behaviour

https://github.com/aio-libs/aiohttp/blob/549279a36a815468a828a98c3375c1f7b2949285/aiohttp/client.py#L569-L571

only the precedence-enforced variant will fail. Nothing random about it :)

We _could_ test for both in one go and remove the parametrization altogether, but splitting the test allows for easier debugging if someone breaks one or both of these behaviours down the line. But again, maybe there's a better way to do this that I'm missing?

Reply:
Ah, I see. I'm not sure if there's a better way, but such unobvious cases always require a code comment explaining these decisions. Otherwise, someone may just refactor this test to drop the parametrization and the if-block because it looks silly :man_shrugging:

Reply:
Very good point, I'll add that in tomorrow with rest of the suggested fixes!